ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -1211,22 +1225,43 @@ void onFillComplete(boolean updateMetric) {
     }
 
     /**
-     * Fill the buffer of the current chunk with the data from the given {@link ByteBuffer}.
-     * @param channelReadBuffer the {@link ByteBuffer} from which to read data.
+     * Fill the buffer of the current chunk with the data from the given {@link ByteBuf}.
+     * @param channelReadBuf the {@link ByteBuf} from which to read data.
      * @return the number of bytes transferred in this operation.
      */
-    int fillFrom(ByteBuffer channelReadBuffer) {
-      int toWrite = Math.min(channelReadBuffer.remaining(), buf.remaining());
-      if (channelReadBuffer.remaining() > buf.remaining()) {
-        // Manipulate limit of the source buffer in order to read only enough to fill the chunk
-        int savedLimit = channelReadBuffer.limit();
-        channelReadBuffer.limit(channelReadBuffer.position() + buf.remaining());
-        buf.put(channelReadBuffer);
-        channelReadBuffer.limit(savedLimit);
+    int fillFrom(ByteBuf channelReadBuf) {
+      int toWrite = 0;
+      if (buf == null) {
+        if (channelReadBuf.readableBytes() > routerConfig.routerMaxPutChunkSizeBytes) {
+          toWrite = routerConfig.routerMaxPutChunkSizeBytes;
+          buf = channelReadBuf.readRetainedSlice(routerConfig.routerMaxPutChunkSizeBytes);
+        } else {
+          toWrite = channelReadBuf.readableBytes();
+          buf = channelReadBuf.readRetainedSlice(toWrite);
+        }
       } else {
-        buf.put(channelReadBuffer);
+        int remainingSize = routerConfig.routerMaxPutChunkSizeBytes - buf.readableBytes();
+        ByteBuf remainingSlice = null;
+        if (channelReadBuf.readableBytes() > remainingSize) {
+          toWrite = remainingSize;
+          remainingSlice = channelReadBuf.readRetainedSlice(remainingSize);
+        } else {
+          toWrite = channelReadBuf.readableBytes();
+          remainingSlice = channelReadBuf.readRetainedSlice(toWrite);
+        }
+        int size = buf.readableBytes();
+        // buf already has some bytes
+        if (buf instanceof CompositeByteBuf) {
```

Comment:
What if `channelReadBuf` was already a `CompositeByteBuf`? Would it be safer to just initialize the buffer in `prepareForBuilding` instead of use type checking logic here? The only thing that I think would be lost would be if the buffer came with a non-default allocator and the logic about direct vs not, but I don't think either of these will be used in practice unless the composite buffers capacity is expanded or components are squashed together inside.

Reply:
I was under the impression that channelReadBuf is passed to PutOperation from HttpContent and shouldn't be composite bytebuf. Is there cases where it will be composite bytebuf?

Reply:
You're probably right that there are no cases in practice with the chunked HttpContents we use. I just wanted to not make too many assumptions about the input in case something changes down the line

Reply:
If we keep this code as it's, then when the `channelReadBuf` is a `CompositeByteBuf`, then when we do `channelReadBuf.readRetainedSlice`, it returns a unpooled ByteBuf. I suppose the code would still operate, but just not with the optimized memory arranagement.